lib/Runtime/Base/ThreadContext.cpp

```diff
@@ -3551,21 +3598,110 @@ ThreadContext::InvalidatePropertyGuardEntry(const Js::PropertyRecord* propertyRe
+{
+    Assert(entry != nullptr);
+    Assert(type != nullptr);
+
+    if (entry->sharedGuard != nullptr)
+    {
+        Js::PropertyGuard* guard = entry->sharedGuard;
+        if (guard->GetValue() == reinterpret_cast<intptr_t>(type))
+        {
+            if (PHASE_TRACE1(Js::TracePropertyGuardsPhase) || PHASE_VERBOSE_TRACE1(Js::FixedMethodsPhase))
+            {
+                Output::Print(_u("FixedFields: invalidating guard: name: %s, address: 0x%p, value: 0x%p, value address: 0x%p\n"),
+                              propertyRecord->GetBuffer(), guard, guard->GetValue(), guard->GetAddressOfValue());
+                Output::Flush();
+            }
+
+            if (PHASE_TESTTRACE1(Js::TracePropertyGuardsPhase) || PHASE_VERBOSE_TESTTRACE1(Js::FixedMethodsPhase))
+            {
+                Output::Print(_u("FixedFields: invalidating guard: name: %s, value: 0x%p\n"), propertyRecord->GetBuffer(), guard->GetValue());
+                Output::Flush();
+            }
+
+            guard->Invalidate();
+        }
+    }
+
+    uint count = 0;
+    entry->uniqueGuards.MapAndRemoveIf([&count, propertyRecord, type](RecyclerWeakReference<Js::PropertyGuard>* guardWeakRef) -> bool
+    {
+        Js::PropertyGuard* guard = guardWeakRef->Get();
+        if (guard != nullptr && guard->GetValue() == reinterpret_cast<intptr_t>(type))
         {
-            // TODO: In future, we should compact the uniqueGuards dictionary so this function can be called from PreCollectionCallback
-            // instead
-            this->ClearInvalidatedUniqueGuards();
-            this->recyclableData->constructorCacheInvalidationCount = 0;
+            if (PHASE_TRACE1(Js::TracePropertyGuardsPhase) || PHASE_VERBOSE_TRACE1(Js::FixedMethodsPhase))
+            {
+                Output::Print(_u("FixedFields: invalidating guard: name: %s, address: 0x%p, value: 0x%p, value address: 0x%p\n"),
+                    propertyRecord->GetBuffer(), guard, guard->GetValue(), guard->GetAddressOfValue());
+                Output::Flush();
+            }
+
+            if (PHASE_TESTTRACE1(Js::TracePropertyGuardsPhase) || PHASE_VERBOSE_TESTTRACE1(Js::FixedMethodsPhase))
+            {
+                Output::Print(_u("FixedFields: invalidating guard: name: %s, value: 0x%p\n"),
+                    propertyRecord->GetBuffer(), guard->GetValue());
+                Output::Flush();
+            }
+
+            guard->Invalidate();
+            count++;
+            return true;
+        }
+        else
+        {
+            return false;
         }
+    });
+
+    if (!isAllPropertyGuardsInvalidation)
+    {
+        this->UpdateConstructorCacheInvalidationCount(count);
     }
 
+    this->UpdateEntryPoints(propertyRecord, entry);
+}
+
+void ThreadContext::UpdateConstructorCacheInvalidationCount(uint count)
```

Comment:
This is a pretty lousy name for this method, but I'm not all that sure what it does.  If you have suggestions for a better name, please let me know!